File: android/measure/src/main/java/sh/measure/android/layoutinspector/LayoutSnapshotThrottler.kt
package sh.measure.android.layoutinspector

import sh.measure.android.utils.TimeProvider
import java.util.concurrent.atomic.AtomicLong

/**
 * Controls the frequency of layout snapshots by enforcing a minimum time interval between captures.
 */
internal class LayoutSnapshotThrottler(private val timeProvider: TimeProvider) {
    private val lastSnapshotAttemptTimestamp = AtomicLong(0)

    /**
     * Determines whether a new snapshot should be taken based on the elapsed time since the
     * last attempt.
     *
     * @param delayMs The required delay between snapshots in milliseconds.
     * @return `true` if enough time has elapsed since the last snapshot or if this is the
     *          first snapshot, false` if the delay hasn't been met.
     */
    fun shouldTakeSnapshot(delayMs: Int = 750): Boolean {
        val now = timeProvider.now()
        val previous = lastSnapshotAttemptTimestamp.getAndSet(now)
        return previous == 0L || (now - previous) > delayMs
    }
}

File: flutter/packages/measure_flutter/lib/src/gestures/layout_snapshot_throttler.dart
import 'package:measure_flutter/src/time/time_provider.dart';

/// Limits how often layout snapshots are taken by requiring a minimum time
/// between them. Every call counts as an attempt, so taps closer together than
/// the delay keep only the first snapshot.
class LayoutSnapshotThrottler {
  static const defaultDelayMs = 750;

  final TimeProvider _timeProvider;
  int _lastAttemptTimestamp = 0;

  LayoutSnapshotThrottler(this._timeProvider);

  bool shouldTakeSnapshot({int delayMs = defaultDelayMs}) {
    final now = _timeProvider.now();
    final previous = _lastAttemptTimestamp;
    _lastAttemptTimestamp = now;
    return previous == 0 || (now - previous) > delayMs;
  }
}
